### PR TITLE
[SV][Prepare] Improve a namehint propagation and spilling logic

### PR DESCRIPTION
**Pull request by uenoku**
This implements a name hint propagation to a wire op canonicalization. 
```
%foo = sv.wire
%0 = comb.add ....
sv.assign %foo, %0
...
```
will be converted into
```
%0 = comb.add {sv.namehint = "foo"}
...
```
This also improves wire spilling logic by allowing expressions used only procedural assignments.

### CODE REVIEW

File: lib/Conversion/ExportVerilog/PrepareForEmission.cpp [hunk 1]
```diff
@@ -615,8 +615,7 @@ bool EmittedExpressionStateManager::shouldSpillWireBasedOnState(Operation &op) {
   // If the operation is only used by an assignment, the op is already spilled
   // to a wire.
   if (op.hasOneUse() &&
-      isa<hw::OutputOp, sv::AssignOp, sv::BPAssignOp, sv::PAssignOp>(
```
Comment (mikeurbach) [inline lib/Conversion/ExportVerilog/PrepareForEmission.cpp:618]: Why was `sv::PAssignOp` removed here? Is this maybe an unrelated change that snuck in? I wouldn't think this is required for this change.

File: lib/Dialect/SV/SVOps.cpp [hunk 2]
```diff
@@ -1483,6 +1483,14 @@ LogicalResult WireOp::canonicalize(WireOp wire, PatternRewriter &rewriter) {
     // use-before-def checking to do, so we only handle that for now.
     return failure();
 
+  // If the wire has a name attribute, propagate the name to the expression.
+  if (auto *connectedOp = connected.getDefiningOp()) {
```
Comment (mikeurbach) [inline lib/Dialect/SV/SVOps.cpp:1487]: nit: i think you can elide the curly braces on the outer if.

**Review by mikeurbach — COMMENTED** [review]
I'm not sure about the change to PrepareForEmission, but the update in the wire canonicalizer looks great. I think this is always going to be an improvement, but I do wonder what about if the existing expression already had a namehint. I guess the name of the wire is a "stronger" hint, so I think what you have here looks good.

**Comment by uenoku** [comment]
> I do wonder what about if the existing expression already had a namehint. I guess the name of the wire is a "stronger" hint, so I think what you have here looks good.

Right, currently a name on a wire overwrites expressions's namehint. The intention is that the name on the wire is generally better than a namehint. I agree we might want to introduce a notion of priority of names.

File: lib/Conversion/ExportVerilog/PrepareForEmission.cpp (hunk 1 above)
Comment (uenoku) [reply lib/Conversion/ExportVerilog/PrepareForEmission.cpp:618]: This is unrelated to the PR but it's necessary for name propagation improvement. I'll split a PR.

Comment (mikeurbach) [reply lib/Conversion/ExportVerilog/PrepareForEmission.cpp:618]: It doesn't have to be a separate PR if it's related, I'm just curious why it is necessary here, since we are only updating the canonicalizer or wires.

Comment (uenoku) [reply lib/Conversion/ExportVerilog/PrepareForEmission.cpp:618]: The condition was intended to avoid spilling same expressions twice by skipping expressions that have single use of assignments. Therefore an expression used by procedural assignments was not spilled to a wire. For example, 
```
%0 = comb.add {sv.namehint = "good_name"}
sv.always .. {
   sv.passign %reg, %0 : ..
}
```
In this case `%0` was inlined to the assignment, but we want to create a wire for it.

Comment (mikeurbach) [reply lib/Conversion/ExportVerilog/PrepareForEmission.cpp:618]: I see, so I think this makes sense then. In general, I guess it's all about tradeoffs, so trading off some inlining for better names sounds good to me.
